File: apps/backend-py/app/utils/distance_calculator.py
"""Distance calculation utilities for route planning.

MVP: Uses Haversine formula for straight-line distances.
Future: Integrate with Google Maps Directions API or OpenStreetMap routing.
"""

import logging
import math
from typing import Dict, List, Tuple

logger = logging.getLogger(__name__)


def haversine_distance(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """
    Calculate great-circle distance between two points using Haversine formula.

    Args:
        lat1, lon1: First point coordinates (degrees)
        lat2, lon2: Second point coordinates (degrees)

    Returns:
        Distance in miles
    """
    # Earth radius in miles
    R = 3959.0

    # Convert to radians
    lat1_rad = math.radians(lat1)
    lat2_rad = math.radians(lat2)
    delta_lat = math.radians(lat2 - lat1)
    delta_lon = math.radians(lon2 - lon1)

    # Haversine formula
    a = (
        math.sin(delta_lat / 2) ** 2
        + math.cos(lat1_rad) * math.cos(lat2_rad) * math.sin(delta_lon / 2) ** 2
    )
    c = 2 * math.asin(math.sqrt(a))

    distance = R * c
    return distance


def calculate_distance_matrix(
    stops: List[Dict[str, any]]
) -> Dict[Tuple[str, str], float]:
    """
    Calculate N x N distance matrix for all stops.

    MVP: Uses Haversine (straight-line) distances.
    Future: Use routing API for actual road distances.

    Args:
        stops: List of stop dicts with keys: stop_id, lat, lon

    Returns:
        Dictionary mapping (stop_id_1, stop_id_2) -> distance_miles
    """
    distance_matrix = {}

    for i, stop1 in enumerate(stops):
        for j, stop2 in enumerate(stops):
            if i == j:
                # Same stop
                distance_matrix[(stop1["stop_id"], stop2["stop_id"])] = 0.0
            else:
                # Calculate distance
                distance = haversine_distance(
                    stop1["lat"], stop1["lon"], stop2["lat"], stop2["lon"]
                )

                # Apply road factor (straight-line * 1.2 for road routing approximation)
                distance_with_road_factor = distance * 1.2

                distance_matrix[(stop1["stop_id"], stop2["stop_id"])] = (
                    distance_with_road_factor
                )

    logger.info(f"Calculated distance matrix for {len(stops)} stops")
    return distance_matrix


def estimate_drive_time(distance_miles: float, road_type: str = "highway") -> float:
    """
    Estimate drive time based on distance and road type.

    MVP: Uses hardcoded average speeds.
    Future: Integrate live traffic data.

    Args:
        distance_miles: Distance in miles
        road_type: 'interstate', 'highway', or 'city'

    Returns:
        Drive time in hours
    """
    # Average speeds by road type (mph)
    speeds = {"interstate": 60, "highway": 50, "city": 30}

    avg_speed = speeds.get(road_type, 55)  # Default to 55 mph

    drive_time_hours = distance_miles / avg_speed
    return drive_time_hours


def get_data_source_label() -> str:
    """
    Return data source label for UI display.

    Returns:
        Data source description
    """
    return "Static Haversine Distance (Future: Google Maps API)"

File: apps/backend/src/utils/distance-calculator.ts
import { Logger } from '@nestjs/common';
import axios from 'axios';

const logger = new Logger('DistanceCalculator');

export type DistanceMethod = 'haversine' | 'google_maps';

export interface DistanceResult {
  distanceMiles: number;
  durationHours: number;
  method: DistanceMethod;
}

/**
 * Calculate straight-line distance between two points using Haversine formula
 * Fast, accurate within ~10-15% of road distance
 * Road factor of 1.2 applied for better approximation
 */
export function haversineDistance(lat1: number, lon1: number, lat2: number, lon2: number): number {
  // Earth radius in miles
  const R = 3959.0;

  // Convert to radians
  const lat1Rad = lat1 * (Math.PI / 180);
  const lat2Rad = lat2 * (Math.PI / 180);
  const deltaLat = (lat2 - lat1) * (Math.PI / 180);
  const deltaLon = (lon2 - lon1) * (Math.PI / 180);

  // Haversine formula
  const a =
    Math.sin(deltaLat / 2) * Math.sin(deltaLat / 2) +
    Math.cos(lat1Rad) * Math.cos(lat2Rad) *
    Math.sin(deltaLon / 2) * Math.sin(deltaLon / 2);
  const c = 2 * Math.asin(Math.sqrt(a));

  return R * c;
}

/**
 * Calculate distance and duration using Google Maps Distance Matrix API
 * Requires GOOGLE_MAPS_API_KEY environment variable
 * More accurate than Haversine as it follows actual roads
 */
export async function calculateGoogleMapsDistance(
  originLat: number,
  originLon: number,
  destLat: number,
  destLon: number
): Promise<{ distanceMiles: number; durationHours: number }> {
  const apiKey = process.env.GOOGLE_MAPS_API_KEY;

  if (!apiKey) {
    throw new Error('GOOGLE_MAPS_API_KEY not configured. Use Haversine method or set API key.');
  }

  try {
    const response = await axios.get('https://maps.googleapis.com/maps/api/distancematrix/json', {
      params: {
        origins: `${originLat},${originLon}`,
        destinations: `${destLat},${destLon}`,
        key: apiKey,
        units: 'imperial', // Miles
      },
      timeout: 5000, // 5 second timeout
    });

    if (response.data.status !== 'OK') {
      throw new Error(`Google Maps API error: ${response.data.status}`);
    }

    const element = response.data.rows[0]?.elements[0];
    if (element?.status !== 'OK') {
      throw new Error(`No route found: ${element?.status}`);
    }

    const distanceMeters = element.distance.value;
    const durationSeconds = element.duration.value;

    return {
      distanceMiles: distanceMeters * 0.000621371, // Convert meters to miles
      durationHours: durationSeconds / 3600, // Convert seconds to hours
    };
  } catch (error) {
    if (axios.isAxiosError(error)) {
      throw new Error(`Google Maps API request failed: ${error.message}`);
    }
    throw error;
  }
}

/**
 * Calculate distance between two points with automatic method selection
 * Uses Google Maps if API key is configured, otherwise falls back to Haversine
 *
 * @param lat1 Origin latitude
 * @param lon1 Origin longitude
 * @param lat2 Destination latitude
 * @param lon2 Destination longitude
 * @param preferredMethod Optional: force a specific calculation method
 * @returns Distance result with method used
 */
export async function calculateDistance(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number,
  preferredMethod?: DistanceMethod
): Promise<DistanceResult> {
  const method = preferredMethod || (process.env.DISTANCE_CALCULATION_METHOD as DistanceMethod) || 'haversine';

  // If Google Maps is preferred and API key is available, try it first
  if (method === 'google_maps' && process.env.GOOGLE_MAPS_API_KEY) {
    try {
      const result = await calculateGoogleMapsDistance(lat1, lon1, lat2, lon2);
      logger.log(`Distance calculated via Google Maps: ${result.distanceMiles.toFixed(1)} mi, ${result.durationHours.toFixed(2)} hrs`);
      return {
        ...result,
        method: 'google_maps',
      };
    } catch (error) {
      logger.warn(`Google Maps API failed, falling back to Haversine: ${error}`);
      // Fall through to Haversine
    }
  }

  // Use Haversine formula (default or fallback)
  const straightLineDistance = haversineDistance(lat1, lon1, lat2, lon2);
  const distanceMiles = straightLineDistance * 1.2; // Apply road factor
  const durationHours = estimateDriveTime(distanceMiles);

  logger.debug(`Distance calculated via Haversine: ${distanceMiles.toFixed(1)} mi, ${durationHours.toFixed(2)} hrs`);

  return {
    distanceMiles,
    durationHours,
    method: 'haversine',
  };
}

/**
 * Calculate distance matrix for all stop pairs
 * Uses Haversine with road factor for speed (matrix calculations are O(n²))
 * For individual route planning, use calculateDistance() for Google Maps support
 */
export function calculateDistanceMatrix(
  stops: Array<{ stop_id: string; lat: number; lon: number }>
): Map<string, number> {
  const matrix = new Map<string, number>();

  for (let i = 0; i < stops.length; i++) {
    for (let j = 0; j < stops.length; j++) {
      const stop1 = stops[i];
      const stop2 = stops[j];
      const key = `${stop1.stop_id}:${stop2.stop_id}`;

      if (i === j) {
        matrix.set(key, 0.0);
      } else {
        const distance = haversineDistance(stop1.lat, stop1.lon, stop2.lat, stop2.lon);
        // Apply road factor (straight-line * 1.2 for road routing approximation)
        matrix.set(key, distance * 1.2);
      }
    }
  }

  logger.log(`Calculated distance matrix for ${stops.length} stops (using Haversine with 1.2x road factor)`);
  return matrix;
}

export function estimateDriveTime(distanceMiles: number, roadType = 'highway'): number {
  // Average speeds by road type (mph)
  const speeds: Record<string, number> = { interstate: 60, highway: 50, city: 30 };
  const avgSpeed = speeds[roadType] ?? 55; // Default to 55 mph

  return distanceMiles / avgSpeed;
}
